share/man/man1/lutris.1
.\" Generated with help2man.
.TH LUTRIS "1"
.SH NAME
lutris \- video game preservation platform
.SH DESCRIPTION
.SS "Usage:"
.IP
lutris [OPTION] URI
.PP
Run a game directly by adding the parameter lutris:rungame/game\-identifier.
.br
If several games share the same identifier you can use the numerical ID (displayed when running lutris \fB\-\-list\-games\fR) and add lutris:rungameid/numerical\-id.
.br
To install a game, add lutris:install/game\-identifier.
.SS "Help Options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help options
.TP
\fB\-\-help\-all\fR
Show all help options
.TP
\fB\-\-help\-gapplication\fR
Show GApplication options
.TP
\fB\-\-help\-gtk\fR
Show GTK+ Options
.SS "Application Options:"
.TP
\fB\-v\fR, \fB\-\-version\fR
Print the version of Lutris and exit
.TP
\fB\-d\fR, \fB\-\-debug\fR
Show debug messages
.TP
\fB\-i\fR, \fB\-\-install\fR
Install a game from a yml file
.TP
\fB\-b\fR, \fB\-\-output\-script\fR
Generate a bash script to run a game without the client
.TP
\fB\-e\fR, \fB\-\-exec\fR
Execute a program with the lutris runtime
.TP
\fB\-l\fR, \fB\-\-list\-games\fR
List all games in database
.TP
\fB\-o\fR, \fB\-\-installed\fR
Only list installed games
.TP
\fB\-s\fR, \fB\-\-list\-steam\-games\fR
List available Steam games
.TP
\fB\-\-list\-steam\-folders\fR
List all known Steam library folders
.TP
\fB\-j\fR, \fB\-\-json\fR
Display the list of games in JSON format
.TP
\fB\-\-reinstall\fR
Reinstall game
.TP
\fB\-\-submit\-issue\fR
Submit an issue
.TP
\fB\-\-display\fR=\fI\,DISPLAY\/\fR
X display to use


debian/changelog
lutris (0.5.10) impish; urgency=medium

  * Add new window to add games to Lutris, with searches from the website,
    scanning a folder for previously installed games, installing a Windows
    game from a setup file, installing from a YAML script or configuring a
    single game manually.
  * Move the search for Lutris installers from a tab in the Lutris service
    to the window for adding games.
  * Add a coverart format
  * Add integration with EA Origin
  * Add integration with Ubisoft Connect
  * Download missing media on startup
  * Remove Winesteam runner (install Steam for Windows in Lutris instead)
  * PC (Linux and Windows) games have their own dedicated Nvidia shader cache
  * Add dgvoodoo2 option
  * Add option to enable BattleEye anti-cheat support
  * Default to Retroarch cores in ~/.config/retroarch/cores if available
  * Add support for downloading patches and DLC for GOG games
  * Add --export and --import command line flags to export a game a lutris
    game and re-import it (requires --dest for the destination path)
  * Add command line flags to manage runners: --install-runner, --uninstall-runners,
    --list-runners, --list-wine-versions
  * Change behavior of the "Stop" button, remove "Kill all Wine processes" action
  * Gamescope option is now disabled on Nvidia GPUs
  * Enable F-Sync by default

 -- Mathieu Comandon <strider@strycore.com>  Fri, 04 Feb 2022 19:14:41 -0800

lutris (0.5.9.1) hirsute; urgency=medium

  * Fix possible escaping error for gamescope option
  * Remove walrus operator to restore compatibility with Python 3.7 / Ubuntu 18.04
  * Remove log file being written in the home folder
  * Fix install button for community installer
  * Fix markup error on gamescope option
  * Update URL for Ryujinx build
  * Fix Steam sync creating duplicate games

 -- Mathieu Comandon <strider@strycore.com>  Sat, 16 Oct 2021 18:08:26 -0700

lutris (0.5.9) hirsute; urgency=medium

  * Add initial support for Epic Games Store
  * Add support for Steam for Windows as a game source
  * Add support for DXVK-NVAPI and DLSS
  * Add FidelityFX Super Resolution (FSR) option for compatible Wine versions
  * Add workaround for locale issues when Lutris is launched from Steam
  * Add gamescope option
  * Lutris games can now be launched from Steam
  * 3rd party services can be enabled or disabled in the preferences
  * The main preferences window has now tabs on the left side
  * Runner configuration is now available from the main preferences window
  * VKD3D is a separate option from DXVK
  * Esync is enabled by default
  * Dolphin is available as a game source (reads games from the emulator's
    local database of games)
  * Scan for installed games when using Steam source
  * Improved automatic installers for GOG, detection of DOSBOX and ScummVM
    games.
  * DRM free services (Humble, GOG) can locate existing installations of games
  * Use 7zip as the default extractor when not given an archive type
  * Improve process monitoring, allowing for monitoring of Steam games
  * Disable AMD switchable graphics layer by default (breaks games)
  * Removed support for Gallium 9
  * Removed support for X360CE
  * Removed legacy WineD3D options

 -- Mathieu Comandon <strider@strycore.com>  Mon, 11 Oct 2021 12:33:39 -0700

lutris (0.5.8.3) groovy; urgency=medium

  * Really fix popovers not showing on Wayland without making them non-modal
  * Prevent GStreamer based configuration from being applied in incompatible
    wine builds.
  * Fix crash when wine runner accesses DXVK versions before they are
    uploaded.
  * Prevent init dialog from being closed while it downloads the runtime.


 -- Mathieu Comandon <strider@strycore.com>  Fri, 22 Jan 2021 16:24:51 -0800

lutris (0.5.8.2) groovy; urgency=medium

  * Fix popover menus not appearing on Wayland
  * Fix game bar getting unselected on Wayland (Forces the last game to stay
    selected)
  * Update Chinese, Dutch, German and Russian translations
  * Download DXVK when Lutris starts
  * Add fsync2 feature detection
  * Limit simultaneous downloads to 3
  * Add support for deb file extraction
  * Add support for Adobe Air games from Humble Bundle (Installation only,
    Air runtime will come at a later stage)
  * Add support for GStreamer enabled Wine builds. This will provide better
    compatibility for games using Media Foundation

 -- Mathieu Comandon <strider@strycore.com>  Mon, 04 Jan 2021 23:54:29 -0800

lutris (0.5.8.1) groovy; urgency=medium

  * Remove Proton from available Wine versions
  * Display a dialog until Lutris finishes initializing
  * Allow to keep game files when uninstalling a game
  * Remove custom sidebar CSS
  * Fix popup menu not showing in list view
  * Fix script loading for local files
  * Fix installed at column setting name for list view
  * Fix lutris not launching games with rungameid
  * Fix installed Steam game for fresh lutris installs

 -- Mathieu Comandon <strider@strycore.com>  Fri, 27 Nov 2020 14:23:48 -0800

lutris (0.5.8) groovy; urgency=medium

  * 3rd party services are now available from the main window
  * The "Import games" window has been removed. The concept of
    importing games from other services into Lutris has been removed.
    Syncing games from other services on start has been removed.
  * Integration with the lutris website such as login and showing your
    library has been delegated to the 'lutris' service in the sidebar.
  * The lutris service gives the option of searching your library or the whole
    lutris.net library.
  * Games from 3rd party services no longer depend on an install script to be
    present on the website. Lutris will automatically install games with an
    auto-generated script. Scripts from the website take precedence if available.
  * Steam games are directly loaded from the Steam API and it is no longer needed
    to sync your Steam library on the lutris website to see all your Steam games.

lutris/services/lutris.py
import json
import os
import urllib.parse
from gettext import gettext as _

from gi.repository import Gio

from lutris import settings
from lutris.api import get_api_games, get_game_installers, read_api_key
from lutris.database.games import get_games
from lutris.database.services import ServiceGameCollection
from lutris.gui import dialogs
from lutris.gui.views.media_loader import download_media
from lutris.services.base import LutrisBanner, LutrisCoverart, LutrisCoverartMedium, LutrisIcon, OnlineService
from lutris.services.service_game import ServiceGame
from lutris.util import http
from lutris.util.log import logger


class LutrisGame(ServiceGame):
    """Service game created from the Lutris API"""
    service = "lutris"

    @classmethod
    def new_from_api(cls, api_payload):
        """Create an instance of LutrisGame from the API response"""
        service_game = LutrisGame()
        service_game.appid = api_payload['slug']
        service_game.slug = api_payload['slug']
        service_game.name = api_payload['name']
        service_game.details = json.dumps(api_payload)
        return service_game


class LutrisService(OnlineService):
    """Service for Lutris games"""

    id = "lutris"
    name = _("Lutris")
    icon = "lutris"
    online = True
    medias = {
        "icon": LutrisIcon,
        "banner": LutrisBanner,
        "coverart_med": LutrisCoverartMedium,
        "coverart_big": LutrisCoverart,
    }
    default_format = "banner"

    api_url = settings.SITE_URL + "/api"
    login_url = settings.SITE_URL + "/api/accounts/token"
    cache_path = os.path.join(settings.CACHE_DIR, "lutris")
    token_path = os.path.join(settings.CACHE_DIR, "auth-token")

    is_loading = False

    @property
    def credential_files(self):
        """Return a list of all files used for authentication"""
        return [self.token_path]

    def match_games(self):
        """Matching lutris games is much simpler... No API call needed."""
        service_games = {
            str(game["appid"]): game for game in ServiceGameCollection.get_for_service(self.id)
        }
        for lutris_game in get_games():
            self.match_game(service_games.get(lutris_game["slug"]), lutris_game)

    def is_connected(self):
        """Is the service connected?"""
        return self.is_authenticated()

    def login(self, parent=None):
        """Connect to Lutris"""
        login_dialog = dialogs.ClientLoginDialog(parent=parent)
        login_dialog.connect("connected", self.on_connect_success)

    def on_connect_success(self, _widget, _username):
        """Handles connection success"""
        self.emit("service-login")

    def get_library(self):
        """Return the remote library as a list of dicts."""
        credentials = read_api_key()
        if not credentials:
            return []
        url = settings.SITE_URL + "/api/games/library/%s" % urllib.parse.quote(credentials["username"])
        request = http.Request(url, headers={"Authorization": "Token " + credentials["token"]})
        try:
            response = request.get()
        except http.HTTPError as ex:
            logger.error("Unable to load library: %s", ex)
            return []
        response_data = response.json
        if response_data:
            return response_data["games"]
        return []

    def load(self):
        if self.is_loading:
            logger.warning("Lutris games are already loading")
            return
        self.is_loading = True
        lutris_games = self.get_library()
        logger.debug("Loaded %s games from Lutris library", len(lutris_games))
        for game in lutris_games:
            lutris_game = LutrisGame.new_from_api(game)
            lutris_game.save()
        logger.debug("Matching with already installed games")
        self.match_games()
        self.is_loading = False
        logger.debug("Lutris games loaded")
        return lutris_games

    def install(self, db_game):
        if isinstance(db_game, dict):
            slug = db_game["slug"]
        else:
            slug = db_game
        installers = get_game_installers(slug)
        if not installers:
            logger.warning("No installer for %s", slug)
            return
        application = Gio.Application.get_default()
        application.show_installer_window(installers)


def download_lutris_media(slug):
    """Download all media types for a single lutris game"""
    url = settings.SITE_URL + "/api/games/%s" % slug
    request = http.Request(url)
    try:
        response = request.get()
    except http.HTTPError as ex:
        logger.debug("Unable to load %s: %s", slug, ex)
        return
    response_data = response.json
    icon_url = response_data.get("icon_url")
    if icon_url:
        download_media({slug: icon_url}, LutrisIcon())

    banner_url = response_data.get("banner_url")
    if banner_url:
        download_media({slug: banner_url}, LutrisBanner())

    cover_url = response_data.get("coverart")
    if cover_url:
        download_media({slug: cover_url}, LutrisCoverart())


def sync_media():
    """Downlad all missing media"""
    banners_available = {fn.split(".")[0] for fn in os.listdir(settings.BANNER_PATH)}
    icons_available = {
        fn.split(".")[0].replace("lutris_", "")
        for fn in os.listdir(settings.ICON_PATH)
        if fn.startswith("lutris_")
    }
    covers_available = {fn.split(".")[0] for fn in os.listdir(settings.COVERART_PATH)}
    complete_games = banners_available.intersection(icons_available).intersection(covers_available)
    all_slugs = {game["slug"] for game in get_games()}
    slugs = all_slugs - complete_games
    if not slugs:
        return
    games = get_api_games(list(slugs))

    alias_map = {}
    api_slugs = set()
    for game in games:
        api_slugs.add(game["slug"])
        for alias in game["aliases"]:
            if alias["slug"] in slugs:
                alias_map[game["slug"]] = alias["slug"]
    alias_slugs = set(alias_map.values())
    used_alias_slugs = alias_slugs - api_slugs
    for alias_slug in used_alias_slugs:
        for game in games:
            if alias_slug in [alias["slug"] for alias in game["aliases"]]:
                game["slug"] = alias_map[game["slug"]]
                continue
    banner_urls = {
        game["slug"]: game["banner_url"]
        for game in games
        if game["slug"] not in banners_available and game["banner_url"]
    }
    icon_urls = {
        game["slug"]: game["icon_url"]
        for game in games
        if game["slug"] not in icons_available and game["icon_url"]
    }
    cover_urls = {
        game["slug"]: game["coverart"]
        for game in games
        if game["slug"] not in covers_available and game["coverart"]
    }
    logger.debug(
        "Syncing %s banners, %s icons and %s covers",
        len(banner_urls), len(icon_urls), len(cover_urls)
    )
    download_media(banner_urls, LutrisBanner())
    download_media(icon_urls, LutrisIcon())
    download_media(cover_urls, LutrisCoverart())
